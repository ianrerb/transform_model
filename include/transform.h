#ifndef TRANSFORM_H
#define TRANSFORM_H
#define PI M_PI 
#include"pricemodel.h"
#include<complex.h>
#include<fftw3.h>
#include<math.h>
#include<vector>
#include<iostream>
#include<string>

struct Option {double strike; double premium; }; //struct to hold strike, premium pairs

/********************************************
 * ABSTRACT BASE CLASS FOR TRANSFORM CLASSES 
 ********************************************/

class transform_base {//abc
  public:
    transform_base(unsigned int size, bool test = false):N_(size), testmode(test){ };
    virtual ~transform_base(){ };
    
    void N(const unsigned int &val){ N_ = val; };
    unsigned int N() const { return N_; };

    virtual std::vector<Option> Prices(const double &Spot, const double &Strike, const double &C, const pricemodel &model) const =0;

  protected:
    unsigned int N_;
   
    bool testmode;
    void printComplexArray(double complex array[], unsigned int length, std::string title) const {
      std::cout<<"Array: "<<title<<std::endl;
      for(unsigned int i=0; i!=length; i++){
	std::cout<<"("<<creal(array[i])<<","<<cimag(array[i])<<")"<<"  ";
	}
      std::cout<<std::endl;
      };
 };
 
/**********************************
 * BEGIN DEFINITION OF FFT CLASS 
 **********************************/

 class FFT : public transform_base {
  public:
    FFT(unsigned int size, double alpha_,double eta_, bool test = false) : transform_base(size,test), alpha(alpha_), eta(eta_){ }; 
    virtual ~FFT(){ }; 
    
    void Alpha(const double &val){ alpha = val; };
    void Eta(const double &val){eta = val; };
    
    double Alpha() const { return alpha; };
    double Eta() const { return eta; };
    virtual double Lambda() const {return 2.0*PI/(static_cast<double>(N_)*eta); };
    
    virtual std::vector<Option> Prices(const double &Spot, const double &Strike, const double &C, const pricemodel &model) const; //defined in FFTengine.cpp

  protected:
    double alpha, eta;

    void ComputeX(double complex X[], const double &Spot, const double &Strike, const double &C, const pricemodel &model) const;  //defined in FFTengine.cpp
  };

/**********************************
 * BEGIN DEFINITION OF FrFFT CLASS 
 **********************************/

class FrFFT : public FFT {
  public:
  FrFFT(unsigned int size, double alpha_, double eta_, double lambda_, bool test = false): FFT(size, alpha_, eta_,test), lambda(lambda_){};
  ~FrFFT(){ };
  
  void Lambda(const double &l){ lambda = l; };
  double Lambda() const { return lambda; };
  
  double Gamma() const { return eta*lambda/(2.0*PI); };

  std::vector<Option> Prices(const double &Spot, const double &Strike, const double &C, const pricemodel &model) const; //defined in FrFFT.cpp
  
  private:
    double lambda;
         
    void ComputeY(double complex Y[],const double complex X[]) const; //defined in FrFFT.cpp

    void ComputeZ(double complex Z[]) const; //defined in FrFFT.cpp

    void ComputeXi(double complex Xi[], const double complex Y[], const double complex Z[]) const; //defined in FrFFT.cpp

  };

#endif
