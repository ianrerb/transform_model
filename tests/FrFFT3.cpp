/* This test checks that the accessability of the parameters for the FrFFT class */ 

#include"../include/transform.h"
#include<iostream>
#include<cmath>

using namespace std;

int main(){
  FrFFT test(64,1,.15,.01,true);
  GBM model(.25,.05,0,1);

  test.Prices(100,100,exp(-.05),model);
  return 0;
  }
