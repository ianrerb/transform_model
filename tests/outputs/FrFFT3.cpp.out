
 <++ THIS IS RAW X ARRAY ++>
(419.652 , 6.42386e-15)
(-810.264 , 166.094)
(731.054 , -306.693)
(-620.867 , 406.312)
(500.771 , -462.442)
(-386.667 , 481.755)
(287.321 , -474.467)
(-205.735 , 450.373)
(141.403 , -417.151)
(-92.1647 , 380.137)
(55.3525 , -342.72)
(-28.3906 , 306.879)
(9.04272 , -273.645)
(4.52312 , 243.454)
(-13.7566 , -216.384)
(19.7798 , 192.313)
(-23.449 , -171.015)
(25.4106 , 152.224)
(-26.1496 , -135.664)
(26.0276 , 121.074)
(-25.3128 , -108.209)
(24.2034 , 96.8524)
(-22.8449 , -86.8111)
(21.3435 , 77.9168)
(-19.7759 , -70.023)
(18.1969 , 63.003)
(-16.6452 , -56.7474)
(15.1471 , 51.1619)
(-13.7202 , -46.1651)
(12.3752 , 41.6866)
(-11.1181 , -37.6657)
(9.95149 , 34.0499)
(-8.87514 , -30.7935)
(7.88718 , 27.8568)
(-6.98448 , -25.2051)
(6.16305 , 22.8083)
(-5.4184 , -20.6397)
(4.74573 , 18.676)
(-4.14015 , -16.8966)
(3.59674 , 15.2833)
(-3.11068 , -13.8199)
(2.67732 , 12.4921)
(-2.29218 , -11.287)
(1.95102 , 10.1931)
(-1.64985 , -9.20025)
(1.3849 , 8.29908)
(-1.15267 , -7.48129)
(0.949925 , 6.73935)
(-0.773644 , -6.06646)
(0.621057 , 5.45643)
(-0.489617 , -4.90366)
(0.376993 , 4.40306)
(-0.281056 , -3.94997)
(0.199867 , 3.54017)
(-0.131667 , -3.16979)
(0.074867 , 2.83529)
(-0.0280294 , -2.53347)
(-0.0101371 , 2.26136)
(0.0407909 , -2.01626)
(-0.0649679 , 1.79572)
(0.0835914 , -1.59746)
(-0.0974814 , 1.41944)
(0.107364 , -1.25974)
(-0.11388 , 1.11666)

 <++ THIS IS RAW Y ARRAY ++>
(419.652 , 6.42386e-15)
(-810.139 , 166.702)
(730.131 , -308.885)
(-618.111 , 410.494)
(495.186 , -468.418)
(-377.567 , 488.92)
(274.408 , -482.051)
(-189.049 , 457.628)
(121.225 , -423.455)
(-68.9156 , 385.031)
(29.5169 , -345.904)
(-0.462706 , 308.189)
(-20.5062 , -273.026)
(35.2621 , 240.93)
(-45.3023 , -212.036)
(51.7978 , 186.259)
(-55.6515 , -163.398)
(57.5527 , 143.197)
(-58.0243 , -125.387)
(57.4611 , 109.702)
(-56.1602 , -95.8956)
(54.3446 , 83.7427)
(-52.1812 , -73.0423)
(49.7944 , 63.6167)
(-47.2768 , -55.3101)
(44.697 , 47.9863)
(-42.1054 , -41.5267)
(39.5387 , 35.8283)
(-37.0235 , -30.8012)
(34.5787 , 26.3672)
(-32.2172 , -22.4582)
(29.9478 , 19.0146)
(-27.7761 , -15.9843)
(25.705 , 13.3214)
(-23.7359 , -10.9858)
(21.8689 , 8.94174)
(-20.1028 , -7.15786)
(18.436 , 5.60619)
(-16.8663 , -4.26183)
(15.3912 , 3.10256)
(-14.0079 , -2.10848)
(12.7133 , 1.26171)
(-11.5044 , -0.546172)
(10.378 , -0.0526374)
(-9.33094 , 0.547832)
(8.35988 , -0.95128)
(-7.46163 , 1.27374)
(6.63292 , -1.52498)
(-5.87053 , 1.71385)
(5.17124 , -1.84841)
(-4.53186 , 1.93595)
(3.94922 , -1.9831)
(-3.4202 , 1.99587)
(2.94171 , -1.97967)
(-2.5107 , 1.9394)
(2.12417 , -1.87947)
(-1.77919 , 1.8038)
(1.47289 , -1.71594)
(-1.20244 , 1.61899)
(0.965128 , -1.5157)
(-0.758297 , 1.4085)
(0.579388 , -1.29947)
(-0.425936 , 1.1904)
(0.295576 , -1.08283)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)
(0 , 0)

 <++ THIS IS RAW Z ARRAY ++>
(1 , 0)
(1 , 0.00075)
(0.999996 , 0.003)
(0.999977 , 0.00674995)
(0.999928 , 0.0119997)
(0.999824 , 0.0187489)
(0.999636 , 0.0269967)
(0.999325 , 0.0367417)
(0.998848 , 0.0479816)
(0.998155 , 0.0607126)
(0.997189 , 0.0749297)
(0.995885 , 0.0906255)
(0.994174 , 0.10779)
(0.991978 , 0.126411)
(0.989215 , 0.146471)
(0.985795 , 0.16795)
(0.981625 , 0.190823)
(0.976602 , 0.215057)
(0.97062 , 0.240616)
(0.963571 , 0.267454)
(0.955336 , 0.29552)
(0.945799 , 0.324752)
(0.934836 , 0.35508)
(0.922322 , 0.386423)
(0.90813 , 0.418688)
(0.892134 , 0.451771)
(0.874205 , 0.485557)
(0.854219 , 0.519914)
(0.832052 , 0.554698)
(0.807585 , 0.589751)
(0.780707 , 0.624897)
(0.751311 , 0.659948)
(0.719302 , 0.694698)
(0.684594 , 0.728925)
(0.647117 , 0.762391)
(0.606814 , 0.794844)
(0.563649 , 0.826015)
(0.517602 , 0.855621)
(0.46868 , 0.883368)
(0.416913 , 0.908946)
(0.362358 , 0.932039)
(0.305103 , 0.952319)
(0.245268 , 0.969455)
(0.183009 , 0.983111)
(0.118517 , 0.992952)
(0.0520228 , 0.998646)
(-0.016203 , 0.999869)
(-0.0858479 , 0.996308)
(-0.156557 , 0.987669)
(-0.227932 , 0.973677)
(-0.299534 , 0.954086)
(-0.370877 , 0.928682)
(-0.441441 , 0.89729)
(-0.510661 , 0.859782)
(-0.577941 , 0.816078)
(-0.642651 , 0.766159)
(-0.704135 , 0.710067)
(-0.761714 , 0.647914)
(-0.814695 , 0.579889)
(-0.862381 , 0.50626)
(-0.904072 , 0.42738)
(-0.939083 , 0.343689)
(-0.966751 , 0.25572)
(-0.986444 , 0.164097)
(-0.986444 , 0.164097)
(-0.966751 , 0.25572)
(-0.939083 , 0.343689)
(-0.904072 , 0.42738)
(-0.862381 , 0.50626)
(-0.814695 , 0.579889)
(-0.761714 , 0.647914)
(-0.704135 , 0.710067)
(-0.642651 , 0.766159)
(-0.577941 , 0.816078)
(-0.510661 , 0.859782)
(-0.441441 , 0.89729)
(-0.370877 , 0.928682)
(-0.299534 , 0.954086)
(-0.227932 , 0.973677)
(-0.156557 , 0.987669)
(-0.0858479 , 0.996308)
(-0.016203 , 0.999869)
(0.0520228 , 0.998646)
(0.118517 , 0.992952)
(0.183009 , 0.983111)
(0.245268 , 0.969455)
(0.305103 , 0.952319)
(0.362358 , 0.932039)
(0.416913 , 0.908946)
(0.46868 , 0.883368)
(0.517602 , 0.855621)
(0.563649 , 0.826015)
(0.606814 , 0.794844)
(0.647117 , 0.762391)
(0.684594 , 0.728925)
(0.719302 , 0.694698)
(0.751311 , 0.659948)
(0.780707 , 0.624897)
(0.807585 , 0.589751)
(0.832052 , 0.554698)
(0.854219 , 0.519914)
(0.874205 , 0.485557)
(0.892134 , 0.451771)
(0.90813 , 0.418688)
(0.922322 , 0.386423)
(0.934836 , 0.35508)
(0.945799 , 0.324752)
(0.955336 , 0.29552)
(0.963571 , 0.267454)
(0.97062 , 0.240616)
(0.976602 , 0.215057)
(0.981625 , 0.190823)
(0.985795 , 0.16795)
(0.989215 , 0.146471)
(0.991978 , 0.126411)
(0.994174 , 0.10779)
(0.995885 , 0.0906255)
(0.997189 , 0.0749297)
(0.998155 , 0.0607126)
(0.998848 , 0.0479816)
(0.999325 , 0.0367417)
(0.999636 , 0.0269967)
(0.999824 , 0.0187489)
(0.999928 , 0.0119997)
(0.999977 , 0.00674995)
(0.999996 , 0.003)
(1 , 0.00075)
(1 , 0)

 <++ THIS IS RAW Xi ARRAY ++>
(51.846 , 65.2135)
(56.8532 , -23.2654)
(-21.9109 , -16.0772)
(4.99094 , 10.9486)
(-1.2525 , -5.96292)
(0.360472 , 3.66501)
(-0.0896946 , -2.41145)
(-0.016697 , 1.74741)
(0.0573457 , -1.29174)
(-0.0801451 , 1.01912)
(0.0857932 , -0.802307)
(-0.0914276 , 0.664784)
(0.0890274 , -0.543987)
(-0.0898084 , 0.465166)
(0.0853557 , -0.390637)
(-0.0845587 , 0.341365)
(0.0797499 , -0.291992)
(-0.0784396 , 0.259184)
(0.0738043 , -0.224718)
(-0.0723664 , 0.201797)
(0.0680616 , -0.176754)
(-0.066641 , 0.160119)
(0.0626942 , -0.141342)
(-0.0613421 , 0.128889)
(0.0577356 , -0.11445)
(-0.0564648 , 0.104887)
(0.0531665 , -0.0935535)
(-0.0519743 , 0.0860476)
(0.0489494 , -0.077001)
(-0.0478271 , 0.0710003)
(0.0450431 , -0.0636776)
(-0.04398 , 0.0588036)
(0.0414077 , -0.0528075)
(-0.040393 , 0.0487942)
(0.038007 , -0.0438377)
(-0.0370305 , 0.0404941)
(0.0348088 , -0.0363652)
(-0.0338613 , 0.0335516)
(0.0317849 , -0.0300913)
(-0.0308582 , 0.0277038)
(0.0289105 , -0.0247906)
(-0.0279971 , 0.0227511)
(0.026164 , -0.020291)
(-0.0252571 , 0.0185399)
(0.023526 , -0.0164593)
(-0.0226197 , 0.0149509)
(0.0209796 , -0.0131916)
(-0.0200682 , 0.0118907)
(0.0185094 , -0.0104061)
(-0.0175877 , 0.00928536)
(0.0161015 , -0.00803834)
(-0.0151646 , 0.0070767)
(0.0137431 , -0.00603706)
(-0.0127862 , 0.00521849)
(0.0114223 , -0.00436188)
(-0.0104409 , 0.00367433)
(0.00912795 , -0.00298109)
(-0.00811736 , 0.00241587)
(0.00684908 , -0.00187021)
(-0.00580465 , 0.00142152)
(0.00457517 , -0.0010109)
(-0.0034921 , 0.000675468)
(0.00229569 , -0.000390207)
(-0.00116896 , 0.000167087)
(0 , 0)
(0.00117575 , -0.000109527)
(-0.00232288 , 0.000163311)
(0.00355342 , -0.000155759)
(-0.00468447 , 9.89074e-05)
(0.0059761 , 3.14997e-05)
(-0.00709705 , -0.000198505)
(0.00845673 , 0.000460003)
(-0.00957394 , -0.000738945)
(0.0110095 , 0.00114252)
(-0.0121298 , -0.00153762)
(0.01365 , 0.0020974)
(-0.014781 , -0.00261563)
(0.0163959 , 0.00334946)
(-0.0175461 , -0.00400095)
(0.0192673 , 0.00493115)
(-0.0204463 , -0.00572986)
(0.0222873 , 0.00688425)
(-0.0235065 , -0.00784878)
(0.0254832 , 0.00926203)
(-0.0267558 , -0.0104169)
(0.028887 , 0.0121323)
(-0.0302287 , -0.0135094)
(0.0325371 , 0.0155817)
(-0.0339671 , -0.0172229)
(0.0364802 , 0.0197212)
(-0.0380223 , -0.0216812)
(0.0407737 , 0.0246942)
(-0.0424573 , -0.0270454)
(0.0454893 , 0.030688)
(-0.0473519 , -0.0335271)
(0.0507174 , 0.0379501)
(-0.0528075 , -0.0414077)
(0.0565748 , 0.0468124)
(-0.058956 , -0.0510677)
(0.0632142 , 0.0577274)
(-0.0659719 , -0.063031)
(0.0708401 , 0.0713246)
(-0.0740917 , -0.0780343)
(0.0797329 , 0.0884993)
(-0.0836438 , -0.097139)
(0.0902874 , 0.11056)
(-0.0950983 , -0.121919)
(0.103078 , 0.139477)
(-0.109153 , -0.15479)
(0.118976 , 0.178336)
(-0.126888 , -0.199611)
(0.139366 , 0.232178)
(-0.150077 , -0.262861)
(0.166593 , 0.309721)
(-0.181837 , -0.356116)
(0.204964 , 0.427124)
(-0.2282 , -0.50177)
(0.263347 , 0.617207)
(-0.302542 , -0.748015)
(0.363278 , 0.955536)
(-0.441345 , -1.21535)
(0.572964 , 1.65089)
(-0.78584 , -2.28158)
(1.24019 , 3.46758)
(-2.39174 , -5.604)
(6.54341 , 10.0978)
(-23.3812 , -13.8521)
(55.6432 , -26.027)
